hftbacktest/src/backtest/feemodel.rs

```diff
@@ -0,0 +1,54 @@
+use crate::types::Order;
+
+pub struct FeeArgs<'a> {
+    pub order: &'a Order,
+    pub amount: f64,
+}
+
+pub trait FeeModel {
+    // Calculates the fee amount.
+    fn amount(&self, args: FeeArgs) -> f64;
+}
+
+// fee based on a percentage of the order value, with the rate
+// depending on whether the order is a maker or taker.
+pub struct TradingValueFeeModel {
+    pub maker_fee: f64,
+    pub taker_fee: f64,
+}
+impl FeeModel for TradingValueFeeModel {
+    fn amount(&self, args: FeeArgs) -> f64 {
+        if args.order.maker {
+            self.maker_fee * args.amount
+        } else {
+            self.taker_fee * args.amount
+        }
+    }
+}
+// fee per trading quantity
+pub struct TradingQtyFeeModel {
+    pub per_qty_fee: f64,
+}
+impl FeeModel for TradingQtyFeeModel {
+    fn amount(&self, args: FeeArgs) -> f64 {
+        self.per_qty_fee * args.order.qty
+    }
+}
+
+// fee per trade
+pub struct FlatPerTradeFeeModel {
+    pub flat_fee: f64,
+}
+impl FeeModel for FlatPerTradeFeeModel {
+    fn amount(&self, args: FeeArgs) -> f64 {
+        self.flat_fee
+    }
+}
+
+// different fees based on the direction.
+pub struct DirectionalFeeModel {}
+impl FeeModel for DirectionalFeeModel {
+    fn amount(&self, args: FeeArgs) -> f64 {
+        panic!("Not implemented");
```

Comment:
his fee model is similar to maker and taker fees, where the fee depends on the side of the transaction, such as Buy and Sell.

Reply:
Is it a flat fee based on direction?

Reply:
A typical example of directional fees is stamp duty. It can be unilateral or bilateral and is usually a percentage of the total transaction amount.

Reply:
Can you give an example?

Reply:
Please refer to the case in [India's NSE](
https://www.nseindia.com/invest/first-time-investor-stamp-duty-charges-taxes). Only the buyer pays the stamp duty. Therefore, the trading fee structure is as follows:

* Buy: transaction fee per share and stamp duty based on the trading value.
* Sell: transaction fee per share.